### PR TITLE
Access to WeakHashMap must be synchronized

### PR DESCRIPTION
**Pull request by ztravis**
...if a thread might be changing the map.

It's unfortunate that you're stuck with the thread-unsafe `SimpleDateFormat` here, since it seems like trying to work around it has caused trouble several times.

### CODE REVIEW

**Comment by benfortuna** [comment]
As the dateformat instance is read only I'm not sure we need to synchronize this. Sure it could create two instances but I don't know if that is really a problem is it? The main thing this solves is two threads sharing the same instance, in order to avoid synchronization.

**Comment by benfortuna** [comment]
Ah sorry, just realised it's about modifying the map itself.

**Comment by artheus** [comment]
@ztravis Please explain the potential risk you see in this code without your suggested change. And why this is a required (implied by *must* in title) change.

**Comment by ztravis** [comment]
Sure.  Most of Java's collection classes are not thread-safe and cannot be safely used/modified by multiple threads concurrently without external synchronization.  In particular, `WeakHashMap` has internal state (e.g. backing array, size information, etc.) that may change as new entries are added or existing ones deleted - attempts to read/update the map while these operations are ongoing may lead to errors/undefined behavior due to inconsistencies as this state is updated.  One particular manifestation (which we suffered) is an infinite busy loop, as explained [here](http://mailinator.blogspot.com/2009/06/beautiful-race-condition.html) (for `HashMap.java`, but the implementation/analysis is essentially the same).

The various static `DateFormatCache` objects in `DateTime.java` are susceptible to this, as they each have an (unsynchronized) `WeakHashMap` that is read/updated in `get()`, which is itself called (without synchronization) in several `DateTime` constructors.  This makes the constructors thread-unsafe, so that e.g. constructing instances across several threads could lead to concurrent calls to `threadMap.put` and thence to errors like the busy loop we experienced.

Relevant javadoc snippets:

"Like most collection classes, this class [WeakHashMap] is not synchronized." -[WeakHashMap javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/WeakHashMap.html)

"If multiple threads access a hash map concurrently, and at least one of the threads modifies the map structurally, it _must_ be synchronized externally. (A structural modification is any operation that adds or deletes one or more mappings; merely changing the value associated with a key that an instance already contains is not a structural modification.)" -[HashMap javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html)

**Comment by benfortuna** [comment]
I'm thinking it might be better to have a synchronized block inside the if(null) {..} block, as we only really need to synchronize when adding new entries to the map. Does that make sense?

**Comment by ztravis** [comment]
Unfortunately, even reading from the map is not safe if other threads might be concurrently modifying it - locking only around calls to `put` is not good enough.  As the javadoc says, even if only a single thread is modifying the map structurally, all access to the map must be synchronized.

File: src/main/java/net/fortuna/ical4j/model/DateTime.java [hunk 1]
```diff
@@ -529,7 +529,7 @@ private DateFormatCache(DateFormat dateFormat) {
 			this.templateFormat = dateFormat;
 		}
 
-		public DateFormat get() {
+		public synchronized DateFormat get() {
```
Comment (seohyang25) [inline src/main/java/net/fortuna/ical4j/model/DateTime.java]: @ztravis 
synchronized scope, can be more smaller, like below.

`private final Object getDateFormatLock = new Object();

public DateFormat get() {
	DateFormat dateFormat;
	synchronized (getDateFormatLock) {
		dateFormat = (DateFormat) threadMap.get(Thread.currentThread());
	}
	...
	return dateFormat;
}`

**Comment by seohyang25** [comment]
@benfortuna  @artheus  @ztravis 

In our service using ical4j, we are experiencing a 'Thread-hang' phenomenon which is assumed to 'WeakHashMap' has problem.

- normal condition tomcat JVM status
image

- abnormal condition tomcat JVM status
CPU usage is high
one thread is continuously RUNNING status.
image

- abnormal condition tomcat JVM thread dump
image

- conclusion
we suspect WeakHashMap.get has problem.
so after googling, find is pull-request.

plz, fix-it, and save our developer's monitoring job ^^.

File: src/main/java/net/fortuna/ical4j/model/DateTime.java (hunk 1 above)
Comment (ztravis) [inline src/main/java/net/fortuna/ical4j/model/DateTime.java]: The call to `threadMap.put(Thread.currentThread(), dateFormat)` would also have to be called while synchronized on the same lock.  The only thing you can reasonably pull out of the synchronized scope is the call to `templateFormat.clone()` and I don't think there's much benefit there.

**Comment by ztravis** [comment]
@fatkudu that fix would be fine as well.  Given how the map is used here they are functionally equivalent.

**Comment by seohyang25** [comment]
@ztravis @benfortuna 
Please review, about ConcurrentHashMap.
`private final Map threadMap = new ConcurrentHashMap();`

I make test cases of four version.

- ASIS
going to hang state.
```
public DateFormat get() {
    DateFormat dateFormat = (DateFormat) threadMap.get(Thread.currentThread());
    if (dateFormat == null) {
        dateFormat = (DateFormat) templateFormat.clone();
        threadMap.put(Thread.currentThread(), dateFormat);
    }
    return dateFormat;
}
```


- by ztravis, method Synchronized
work well. but some slow.
```
public synchronized DateFormat get() {
    DateFormat dateFormat = (DateFormat) threadMap.get(Thread.currentThread());
    if (dateFormat == null) {
        dateFormat = (DateFormat) templateFormat.clone();
        threadMap.put(Thread.currentThread(), dateFormat);
    }
    return dateFormat;
}
```


- by seohyang25, block lock
going to hang.. not works TT
```
private static final Object getDateFormatLock = new Object();
public DateFormat get() {
    DateFormat dateFormat;
    synchronized (getDateFormatLock) {
        dateFormat = (DateFormat) threadMap.get(Thread.currentThread());
    }
    if (dateFormat == null) {
        dateFormat = (DateFormat) templateFormat.clone();
        threadMap.put(Thread.currentThread(), dateFormat);
    }
    return dateFormat;
}
```


- by CacheWithConcurrentHashMap
good works! Java LOL
```
private static class DateFormatCacheWithConcurrentHashMap implements Testable {
    private final Map threadMap = new ConcurrentHashMap();
    private final DateFormat templateFormat;

    private DateFormatCacheWithConcurrentHashMap(DateFormat dateFormat) {
        this.templateFormat = dateFormat;
    }

    public DateFormat get() {
        DateFormat dateFormat = (DateFormat) threadMap.get(Thread
                .currentThread());
        if (dateFormat == null) {
            dateFormat = (DateFormat) templateFormat.clone();
            threadMap.put(Thread.currentThread(), dateFormat);
        }
        return dateFormat;
    }

    public void doTest() {
        get();
    }
}
```


- Attach my full code.
```
import net.fortuna.ical4j.util.TimeZones;
import org.junit.Test;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.WeakHashMap;
import java.util.concurrent.ConcurrentHashMap;

public class TestOfMap {
    private static final DateFormatCache dateFormatCache;
    private static final DateFormatCacheWithMethodSynchronized dateFormatCacheWithMethodSynchronized;
    private static final DateFormatCacheWithBlockSynchronized dateFormatCacheWithBlockSynchronized;
    private static final DateFormatCacheWithConcurrentHashMap dateFormatCacheWithConcurrentHashMap;

    static {
        final DateFormat format = new SimpleDateFormat("yyyyMMdd'T'HHmmss'Z'");
        format.setTimeZone(TimeZones.getUtcTimeZone());
        format.setLenient(false);

        dateFormatCache = new DateFormatCache(format);
        dateFormatCacheWithMethodSynchronized = new DateFormatCacheWithMethodSynchronized(format);
        dateFormatCacheWithBlockSynchronized = new DateFormatCacheWithBlockSynchronized(format);
        dateFormatCacheWithConcurrentHashMap = new DateFormatCacheWithConcurrentHashMap(format);
    }

    interface Testable {
        void doTest();
    }

    // ASIS
    private static class DateFormatCache implements Testable {
        private final Map threadMap = new WeakHashMap();
        private final DateFormat templateFormat;

        private DateFormatCache(DateFormat dateFormat) {
            this.templateFormat = dateFormat;
        }

        public DateFormat get() {
            DateFormat dateFormat = (DateFormat) threadMap.get(Thread.currentThread());
            if (dateFormat == null) {
                dateFormat = (DateFormat) templateFormat.clone();
                threadMap.put(Thread.currentThread(), dateFormat);
            }
            return dateFormat;
        }

        public void doTest() {
            get();
        }
    }

    // Modify by method synchronized
    private static class DateFormatCacheWithMethodSynchronized implements Testable {
        private final Map threadMap = new WeakHashMap();
        private final DateFormat templateFormat;

        private DateFormatCacheWithMethodSynchronized(DateFormat dateFormat) {
            this.templateFormat = dateFormat;
        }

        public synchronized DateFormat get() {
            DateFormat dateFormat = (DateFormat) threadMap.get(Thread.currentThread());
            if (dateFormat == null) {
                dateFormat = (DateFormat) templateFormat.clone();
                threadMap.put(Thread.currentThread(), dateFormat);
            }
            return dateFormat;
        }

        public void doTest() {
            get();
        }
    }

    // Modify by block synchronized
    private static class DateFormatCacheWithBlockSynchronized implements Testable {
        private final Map threadMap = new WeakHashMap();
        private final DateFormat templateFormat;

        private DateFormatCacheWithBlockSynchronized(DateFormat dateFormat) {
            this.templateFormat = dateFormat;
        }

        private static final Object getDateFormatLock = new Object();

        public DateFormat get() {
            DateFormat dateFormat;
            synchronized (getDateFormatLock) {
                dateFormat = (DateFormat) threadMap.get(Thread.currentThread());
            }
            if (dateFormat == null) {
                dateFormat = (DateFormat) templateFormat.clone();
                threadMap.put(Thread.currentThread(), dateFormat);
            }
            return dateFormat;
        }

        public void doTest() {
            get();
        }
    }

    // Modify by ConcurrentHashMap
    private static class DateFormatCacheWithConcurrentHashMap implements Testable {
        private final Map threadMap = new ConcurrentHashMap();
        private final DateFormat templateFormat;

        private DateFormatCacheWithConcurrentHashMap(DateFormat dateFormat) {
            this.templateFormat = dateFormat;
        }

        public DateFormat get() {
            DateFormat dateFormat = (DateFormat) threadMap.get(Thread
                    .currentThread());
            if (dateFormat == null) {
                dateFormat = (DateFormat) templateFormat.clone();
                threadMap.put(Thread.currentThread(), dateFormat);
            }
            return dateFormat;
        }

        public void doTest() {
            get();
        }
    }

    private class TestRunner extends Thread {
        String runnerName;
        int jobTargetCount;
        Testable testTarget;
        boolean complete = false;

        TestRunner(String runnerName, int jobTargetCount, Testable testTarget) {
            this.runnerName = runnerName;
            this.jobTargetCount = jobTargetCount;
            this.testTarget = testTarget;
        }

        public void run() {
            int count = 0;
            for (int i = 0; i < jobTargetCount; i++) {
                testTarget.doTest();
            }
            complete = true;
            //System.out.println("job done." + runnerName);
        }

        boolean isComplete() {
            return complete;
        }
    }

    private void iterationTest(Testable testTarget) throws InterruptedException {
        int threadCount = 2048;
        int jobTargetCountPerThread = 10000;
        List<TestRunner> threadList = new ArrayList<>();

        for (int iteration = 0; iteration < 50; iteration++) {
            System.out.println("start iteration no." + iteration);
            threadList = new ArrayList<>();
            for (int i = 0; i < threadCount; i++) {
                threadList.add(new TestRunner("runner No." + i, jobTargetCountPerThread, testTarget));
            }
            for (TestRunner thread : threadList) {
                thread.start();
            }

            while (isRemainJob(threadList)) {
                Thread.sleep(5);
            }

            System.out.println("Done  iteration no." + iteration);
        }
    }

    private boolean isRemainJob(List<TestRunner> threadList) {
        for (TestRunner thread : threadList) {
            if (false == thread.isComplete()) {
                return true;
            }
        }
        return false;
    }

    @Test
    public void test_dateFormatCache() throws InterruptedException {
        iterationTest(dateFormatCache);
    }

    @Test
    public void test_methodSynchronized() throws InterruptedException {
        iterationTest(dateFormatCacheWithMethodSynchronized);
    }

    @Test
    public void test_blockSynchronized() throws InterruptedException {
        iterationTest(dateFormatCacheWithBlockSynchronized);
    }

    @Test
    public void test_concurrentHashMap() throws InterruptedException {
        iterationTest(dateFormatCacheWithConcurrentHashMap);
    }
}
```

**Comment by fatkudu** [comment]
How important is it that the map be a `WeakHashMap`?  i.e. drop entries that are no longer reachable because they are garbage collected.
I think that if an application creates lots of short lived threads which use this, the map will leak forever (because it will be retain references to entries no longer used elsewhere)  unless it is a WeakHashMap.  Note that https://docs.oracle.com/javase/7/docs/api/java/util/WeakHashMap.html specifically says "A synchronized WeakHashMap may be constructed using the Collections.synchronizedMap method".
It looks like another possibility if you can use Guava is CacheBuilder.newBuilder().weakKeys().build() - see http://stackoverflow.com/questions/2255950/is-there-java-util-concurrent-equivalent-for-weakhashmap

**Comment by seohyang25** [comment]
@fatkudu 
Thanks. I was only obsessed with "lock" and overlooked the intention of originally using "WeakHashMap". 
After I made the test-code, I proved your advice is right.

```
private static class DateFormatCacheWithSynchronizedMap implements Testable {
    private final Map threadMap = Collections.synchronizedMap(new WeakHashMap<Thread, DateFormat>());
    ...
    public int getMapSize() {
        return threadMap.size();
    }
}
```


- ConcurrentHashMap Test
```
start iteration no.0
Done  iteration no.0 mapSize=2048 diff=2048
start iteration no.1
Done  iteration no.1 mapSize=4096 diff=2048
...
Done  iteration no.49 mapSize=102400 diff=2048 -> just 2048 x 50
Elapsed time : 9667msec
```


- methodSynchronized Test
```
start iteration no.0
Done  iteration no.0 mapSize=2048 diff=2048
...
Done  iteration no.6 mapSize=7099 diff=-5189      -> decrease !
...
start iteration no.49
Done  iteration no.49 mapSize=52610 diff=2048 -> under 2048 x 50
Elapsed time : 50806msec -> slow...
```

- SynchronizedMap Test
```
start iteration no.0
Done  iteration no.0 mapSize=2048 diff=2048
...
start iteration no.49
Done  iteration no.49 mapSize=52414 diff=2048 -> under 2048 x 50
Elapsed time : 47816msec -> faster than method Synchronized
```

**Comment by benfortuna** [comment]
@seohyang25 Would it be possible for you to fork the main repo, apply these changes and submit as another pull request? Otherwise we can find another way I guess..

It looks like you've don't some thorough testing which is really awesome. thanks.

**Comment by ztravis** [comment]
Yes, it's important for this to be a `WeakHashMap`.  Wrapping the map with `Collections.synchronizedMap` is equivalent to wrapping all method invocations on `mapInstance` with `synchronized (mapInstance) {...}`.  I think the important thing here is clarity/commenting so that future developers don't inadvertently break this code again (e.g. the suggestion not to use a `WeakHashMap`, accessing/modifying the map in different places without synchronization).  As I noted above, this small piece of functionality has clearly been a hotspot for bugs and issues in the past (based on previous patches/fixes) so hopefully we can set it straight. I'm happy to make the proposed change and add comments in this PR, if that's ok.

**Comment by benfortuna** [comment]
@ztravis that would be great. I think it's clear we need to replace this approach entirely by using thread-safe date formatting, but that will probably come with the migration to the new date/time api. As a tactical solution I think this sounds ok.

**Comment by douglm** [comment]
Before getting too complicated - and accepting that this is a short term 
fix till we move to Java 8+...

What's the *relative* cost of just creating a new instance or simply 
synching around the uses?

There are only 6-7 uses of the formats and - apart from the option of 
passing your own - there are rather fewer formats.

I'd propose changing it to an explicit formatDate method with a 
parameter indicating which format - if you want to get a little more 
clever allow it to be overridden.

**Comment by benfortuna** [comment]
It's a fair point Mike; I feel like the original reason for the shared dateformat instances was for performance reasons (i.e. avoid duplicating the same format instance for each date/date-time), but perhaps it's just a performance hit we deal with for now for the sake of simplicity..

**Comment by ztravis** [comment]
`SimpleDateFormat` is a rather heavyweight object, so making a new instance every time it's needed is probably not the best option.  I found this (older, informal) reference: http://moradanen.sopovs.com/2012/01/synchronized-vs-threadlocal.html
Another possibility is to switch to Apache Commons lang `FastDateFormat` (until the switch to Java 8).

**Comment by benfortuna** [comment]
We already use apache commons-lang, so if it's compatible then implementing FastDateFormat shouldn't be difficult.

**Comment by ztravis** [comment]
For now I've only made the synchronized changes (switching to `FastDateFormat` might involve some behavioral changes since it doesn't support the `lenient` flag).

**Pull request merged**